src/lib/core/data-table/data-table.ts

```diff
@@ -112,21 +116,32 @@ export class CdkTable implements CollectionViewer {
     this._columnDefinitions.forEach(columnDef => {
       this._columnDefinitionsByName.set(columnDef.name, columnDef);
     });
+
+    // Get and merge the streams for column changes made to the row defs
+    const rowDefs = this._rowDefinitions.toArray().concat(this._headerDefinition);
+    const columnChangeStreams = rowDefs.map((rowDef: BaseRowDef) => rowDef.columnsChange);
+    this.columnsChange = Observable.combineLatest(columnChangeStreams);
   }
 
   ngAfterViewInit() {
-    // TODO(andrewseguin): Re-render the header when the header's columns change.
     this.renderHeaderRow();
 
-    // TODO(andrewseguin): Re-render rows when their list of columns change.
+    // Re-render the header row if the columns changed.
+    this.columnsChange.subscribe(() => {
+      this._headerRowPlaceholder.viewContainer.clear();
+      this.renderHeaderRow();
+    });
+
     // TODO(andrewseguin): If the data source is not
     //   present after view init, connect it when it is defined.
     // TODO(andrewseguin): Unsubscribe from this on destroy.
-    this.dataSource.connect(this).subscribe((rowsData: any[]) => {
+    const streams = [this.dataSource.connect(this), this.columnsChange];
+    Observable.combineLatest(streams).subscribe((result: any[]) => {
```

Comment:
I don't think it can be since I cannot type the individual elements of the array.